Relocate existing k8s/terraform infrastructure

brews opened the pull request:
This PR migrates two bits of existing infrastructure. Moving this logic here will make our dependence on complex infrastructure stuff more transparent and open.

The first addition is the Kubernetes infrastructure from https://github.com/ClimateImpactLab/downscaleCMIP6-infra-argo. This is moved to a new `/infrastructure/kubernetes/` directory in this repo.

This infra is automatically watched/managed/updated by an ArgoCD install on the cluster.

A README.md at the base of that directory describes what's what and should be readable by anyone familiar with Kubernetes and basic ArgoCD.

The other addition is the terraform deploying. This is migrated from https://github.com/ClimateImpactLab/downscaleCMIP6-infra to `/infrastructure/terraform/` in this repo. This code sets up the hard hardware, including the clusters deployed to in `/infrastructure/kubernetes/`.

Changes to Terraform will be planned automatically through CI/CD, though viewing the plans and applying changes requires manual, external confirmation.

brews commented:
I just added the Terraform infra. I'll need to try to set up CI/CD for it after this merges.

dgergel commented on workflows/notes.sh:
```diff
@@ -0,0 +1,520 @@
+#     obj = cls.from_string(defstr)
+#   File "/opt/conda/lib/python3.8/site-packages/xclim/sdba/base.py", line 82, in from_string
+#     params[dd["name"]] = cls._parametrizable_classes[
+# KeyError: 'Grouper'
```
do we need this bit of commented out code?

brews replied:
Ah, no. Thanks for catching that. The entire file should be removed.

(done)

dgergel commented on workflows/notes.sh:
```diff
@@ -0,0 +1,520 @@
+# dc6-dev-plrg4-580946482:     _init_array_metadata(store, shape=shape, chunks=chunks, dtype=dtype,
+# dc6-dev-plrg4-580946482:   File "/opt/conda/lib/python3.8/site-packages/zarr/storage.py", line 377, in _init_array_metadata
+# dc6-dev-plrg4-580946482:     raise ContainsArrayError(path)
+# dc6-dev-plrg4-580946482: zarr.errors.ContainsArrayError: path 'lat' contains an array
```
should we remove this too?

brews replied:
Yeah, thanks. I'm removing this entire file.

Marking resolved.

dgergel reviewed (changes requested):
@brews this looks so awesome, thank you for setting this all up and migrating it to this repo! I have a few comments, and I was wondering if we want to take out some of the errors that are commented out but appear in the `notes` section? Or did you want to have a record of these? Also was wondering what the point was of this file `infrastructure/terraform/outputs.tf` since it's empty? I'm super green with terraform though so maybe that has a purpose I'm not aware of...

brews commented on infrastructure/kubernetes/README.md:
```diff
@@ -0,0 +1,90 @@
+# downscaleCMIP6-infra-argo
+Kuberenetes internals and Argo Workflow configuration for CMIP6 downscaling project.
```
Got it. Thanks!

Marking resolved.

brews commented on infrastructure/kubernetes/README.md:
```diff
@@ -0,0 +1,90 @@
+
+## Updating
+
+If there is already has a deployment -- file a pull request with changes to `main`. Merged PRs are checked and automatically pulled.
```
I updated it to make it more clear (and grammatically correct) — I hope.

Please let me know if you'd like further changes, @dgergel.

brews commented:
> @brews this looks so awesome, thank you for setting this all up and migrating it to this repo! I have a few comments, and I was wondering if we want to take out some of the errors that are commented out but appear in the `notes` section? Or did you want to have a record of these? Also was wondering what the point was of this file `infrastructure/terraform/outputs.tf` since it's empty? I'm super green with terraform though so maybe that has a purpose I'm not aware of...

Thanks for the feedback, @dgergel. This is ready for a re-review.

The `notes.sh` file is old scratch notes. It shouldn't be here. I removed it.

The outputs.tf is for denoting output variables from a Terrafrom workspace. I removed it. I can drop it back in here if we have need for it again in the future.

Please let me know if you have additional comments or changes. Thank you!

dgergel reviewed (approved):
Looks great, thanks @brews!
